Source/ABLASTR/DepositCharge.H

```diff
@@ -0,0 +1,182 @@
+ * \param load_balance_costs_update_algo selected method for updating load balance costs.
+ * \param do_device_synchronize call amrex::Gpu::synchronize() for tiny profiler regions
+ */
+template <typename PC>
+void DepositCharge (typename PC::ParIterType& pti,
+                    typename PC::RealVector& wp,
+                    const int * const ion_lev,
+                    amrex::MultiFab* rho, const int icomp, const int nc,
+                    const long offset, const long np_to_depose,
+                    amrex::FArrayBox& local_rho, const int lev, const int depos_lev,
+                    const amrex::Real charge, const int nox, const int noy, const int noz,
+                    const amrex::IntVect& ng_rho, const std::array<amrex::Real,3>& dx,
+                    const std::array<amrex::Real, 3>& xyzmin,
+                    const amrex::IntVect& ref_ratio,
+                    amrex::Real* cost, const int n_rz_azimuthal_modes,
+                    const long load_balance_costs_update_algo,
+                    const bool do_device_synchronize)
+{
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE((depos_lev==(lev-1)) ||
+                                     (depos_lev==(lev  )),
+                                     "Deposition buffers only work for lev-1");
+
+    // If no particles, do not do anything
+    if (np_to_depose == 0) return;
+
+    // Extract deposition order and check that particles shape fits within the guard cells.
+    // NOTE: In specific situations where the staggering of rho and the charge deposition algorithm
+    // are not trivial, this check might be too strict and we might need to relax it, as currently
+    // done for the current deposition.
+
+#if   defined(WARPX_DIM_1D_Z)
+    amrex::ignore_unused(nox);
+    amrex::ignore_unused(noy);
+    const amrex::IntVect shape_extent = amrex::IntVect(static_cast<int>(noz/2+1));
+#elif   defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+    amrex::ignore_unused(noy);
+    const amrex::IntVect shape_extent = amrex::IntVect(static_cast<int>(nox/2+1),
+                                                       static_cast<int>(noz/2+1));
+#elif defined(WARPX_DIM_3D)
+    const amrex::IntVect shape_extent = amrex::IntVect(static_cast<int>(nox/2+1),
+                                                       static_cast<int>(noy/2+1),
+                                                       static_cast<int>(noz/2+1));
+#endif
+
+    // On CPU: particles deposit on tile arrays, which have a small number of guard cells ng_rho
+    // On GPU: particles deposit directly on the rho array, which usually have a larger number of guard cells
+#ifndef AMREX_USE_GPU
+    const amrex::IntVect range = ng_rho - shape_extent;
+#else
+    const amrex::IntVect range = rho->nGrowVect() - shape_extent;
+#endif
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+        amrex::numParticlesOutOfRange(pti, range) == 0,
+        "Particles shape does not fit within tile (CPU) or guard cells (GPU) used for charge deposition");
```

Comment:
Is it really worth unconditionally doing an extra pass over all of the particles for this error check? Could the same check be made an assertion inside the kernel that loops over the particles anyway?

Reply:
I tend to agree with you that this is too expensive to always do - FTR this was already the behavior of DepositCharge, not new to this PR. Maybe this is was useful when developing the large time step algorithm, though, @RemiLehe?

Reply:
It's indeed not free but also the first line where we see issues when implementing new algorithms or (over-)optimizing our comms.

Note that we cannot afford full debug builds in CI atm., so we might not catch a new issues immediately if that one is debug only. But (debug || a runtime option) or so would work.

Can we potentially estimate/measure the true cost here?

Reply:
`AMREX_ASSERT_WITH_MESSAGE` will assert if `USE_ASSERTION=TRUE`, even if `DEBUG=FALSE`, so yes I think we should make this `AMREX_ASSERT_WITH_MESSAGE`.

Reply:
OK, yes, I agree that this assertion should only be done in `DEBUG` mode, for performance reasons.

Reply:
PR to do this here: https://github.com/ECP-WarpX/WarpX/pull/2688